### DISCUSSION TITLE
CLI tool & correctly stopping it?

### DISCUSSION THREAD

**Discussion by ringods** [ergo-services/ergo#201] [Q&A] [answered]
I am designing a data migration tool. To perform migration and intermediate transformation steps, each could be very elegantly be implmented as an actor. But when all the data is migrated, how should/would I signal this to all the actors to stop and be able to return an exit code from my main function?

**Comment by halturin** [comment]
I would use `Events` for that https://docs.ergo.services/basics/events

PS I'm working on `act.Saga` now, which is intended specifically for such cases. It will be a part of 3.1 release.

**Reply by ringods** [reply]
What is the 3.1 timeframe?

**Reply by halturin** [reply]
~ April 2025

**Comment by ringods** [comment]
@halturin I will have a look at `Events`. But when I look at the examples, they all seem to end with a `node.Wait()`, for instance:

https://github.com/ergo-services/examples/blob/0cc99885eef07c013ab86d968e7bd123f608c745/call/demo.go#L54

How should I implement a grafeful exit, preferrably with the ability to return an exit code as most CLI's have.

**Reply by halturin** [reply]
Use `Stop` method of `gen.Node` interface  https://docs.ergo.services/basics/node#node-shutdown

There is no way to specify exit code so far. Probably good to add it at some point.

**Reply by ringods** [reply]
@halturin In one actor, I emit an event, with the last event having another message type. When that specific message is received on the subscriber side, I trigger:

```go
	case backend.NoMoreStacksMessage:
		// All stack were emitted, so we stop the CLI
		a.Node().Stop()
```

and I get the following log messages:

```sh
1734687840811712000 [info] <420FB26D.0.1004>: Supervisor: terminate children <420FB26D.0.1005>
1734687840811738000 [info] <420FB26D.0.1004>: Supervisor: terminate children <420FB26D.0.1007>
1734687840811751000 [info] <420FB26D.0.1005>: terminated with reason: <420FB26D.0.1004>: shutdown
```

But my process does not end and return to the shell prompt. Both my Application and Supervisor are configured for transient restarts. What am I missing?

**Reply by halturin** [reply, marked as answer]
missing this one:
![image](https://github.com/user-attachments/assets/f69c105b-6abc-45f7-bc47-ba4f45845228)

**Reply by ringods** [reply]
@halturin thanks for highlighting this. Calling it from a goroutine indeed solved it. Tnx.
